src/sync/sync_permission.cpp

```diff
@@ -0,0 +1,218 @@
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync/sync_permission.hpp"
+
+#include "impl/object_accessor_impl.hpp"
+#include "object_schema.hpp"
+#include "property.hpp"
+
+#include "sync/sync_config.hpp"
+#include "sync/sync_manager.hpp"
+#include "sync/sync_session.hpp"
+#include "sync/sync_user.hpp"
+#include "util/event_loop_signal.hpp"
+#include "util/uuid.hpp"
+
+#include <realm/query_expression.hpp>
+#include <realm/util/to_string.hpp>
+
+using namespace realm;
+
+// MARK: - Utility
+
+namespace {
+
+Permission::AccessLevel extract_access_level(Object& permission, CppContext& context)
+{
+    auto may_manage = permission.get_property_value<util::Any>(&context, "mayManage");
+    if (may_manage.has_value() && any_cast<bool>(may_manage))
+        return Permission::AccessLevel::Admin;
+
+    auto may_write = permission.get_property_value<util::Any>(&context, "mayWrite");
+    if (may_write.has_value() && any_cast<bool>(may_write))
+        return Permission::AccessLevel::Write;
+
+    auto may_read = permission.get_property_value<util::Any>(&context, "mayRead");
+    if (may_read.has_value() && any_cast<bool>(may_read))
+        return Permission::AccessLevel::Read;
+
+    return Permission::AccessLevel::None;
+}
+
+}
+
+// MARK: - PermissionResults
+
+Permission PermissionResults::get(size_t index)
+{
+    Object permission(m_results.get_realm(), m_results.get_object_schema(), m_results.get(index));
+    CppContext context;
+    return {
+        any_cast<std::string>(permission.get_property_value<util::Any>(&context, "path")),
+        extract_access_level(permission, context),
+        { any_cast<std::string>(permission.get_property_value<util::Any>(&context, "userId")) }
+    };
+}
+
+// MARK: - Permissions
+
+// A wrapper that stores a value and an associated notification token.
+// The point of this type is to keep the notification token alive
+// until the value can be properly processed or handled.
+template<typename T>
+struct NotificationWrapper {
+    T object;
+    NotificationToken token;
+};
+
+void Permissions::get_permissions(std::shared_ptr<SyncUser> user,
+                                  PermissionResultsCallback callback,
+                                  const ConfigMaker& make_config)
+{
+    auto realm = Permissions::permission_realm(user, make_config);
+    auto session = SyncManager::shared().get_session(realm->config().path, *realm->config().sync_config);
+    auto results_wrapper = std::make_shared<NotificationWrapper<Results>>();
+    auto table = ObjectStore::table_for_object_type(realm->read_group(), "Permission");
+    results_wrapper->object = Results(std::move(realm), *table);
+
+    auto async = [results_wrapper, callback=std::move(callback)](auto ex) mutable {
+        Results& res = results_wrapper->object;
+        if (ex) {
+            callback(nullptr, ex);
+            results_wrapper.reset();
+        } else if (res.size() > 0) {
```

Comment:
Writing this as an early return would make it the comment about the presence of `__management` easier to follow, IMO.